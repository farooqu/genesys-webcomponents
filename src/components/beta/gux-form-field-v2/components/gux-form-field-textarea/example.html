<h1>gux-form-field-textarea-beta</h1>

<p>slot a textarea element into a gux-form-field-textarea-beta element</p>

<h2>Examples</h2>
<form onchange="notify(event)" oninput="notify(event)">
  <fieldset>
    <legend>Label Position</legend>

    <gux-form-field-textarea-beta>
      <textarea slot="input" name="textarea"></textarea>
      <label slot="label">Default</label>
    </gux-form-field-textarea-beta>

    <gux-form-field-textarea-beta>
      <textarea slot="input" name="textarea"></textarea>
      <label slot="label">Default long label</label>
    </gux-form-field-textarea-beta>

    <gux-form-field-textarea-beta label-position="above">
      <textarea slot="input" name="textarea"></textarea>
      <label slot="label">Above</label>
    </gux-form-field-textarea-beta>

    <gux-form-field-textarea-beta label-position="beside">
      <textarea slot="input" name="textarea"></textarea>
      <label slot="label">Beside</label>
    </gux-form-field-textarea-beta>

    <gux-form-field-textarea-beta label-position="screenreader">
      <textarea slot="input" name="textarea"></textarea>
      <label slot="label">Screenreader</label>
    </gux-form-field-textarea-beta>
  </fieldset>

  <fieldset>
    <legend>Resize</legend>

    <gux-form-field-textarea-beta>
      <textarea slot="input" name="textarea"></textarea>
      <label slot="label">Default</label>
    </gux-form-field-textarea-beta>

    <gux-form-field-textarea-beta resize="auto">
      <textarea slot="input" name="textarea"></textarea>
      <label slot="label">Auto</label>
    </gux-form-field-textarea-beta>

    <gux-form-field-textarea-beta resize="manual">
      <textarea slot="input" name="textarea"></textarea>
      <label slot="label">Manual</label>
    </gux-form-field-textarea-beta>

    <gux-form-field-textarea-beta resize="none">
      <textarea slot="input" name="textarea"></textarea>
      <label slot="label">None</label>
    </gux-form-field-textarea-beta>
  </fieldset>

  <fieldset>
    <legend>Input attributes</legend>

    <gux-form-field-textarea-beta>
      <textarea slot="input" name="textarea" disabled></textarea>
      <label slot="label">Disabled</label>
    </gux-form-field-textarea-beta>

    <gux-form-field-textarea-beta>
      <textarea slot="input" name="textarea" required></textarea>
      <label slot="label">Required</label>
    </gux-form-field-textarea-beta>

    <gux-form-field-textarea-beta>
      <textarea
        slot="input"
        name="textarea"
        placeholder="Enter text"
      ></textarea>
      <label slot="label">Placeholder</label>
    </gux-form-field-textarea-beta>
  </fieldset>

  <fieldset>
    <legend>Error</legend>

    <gux-form-field-textarea-beta>
      <textarea slot="input" name="textarea"></textarea>
      <label slot="label">Default</label>
      <span slot="error">This is an error message</span>
    </gux-form-field-textarea-beta>
  </fieldset>
</form>

<h2>gux-form-field examples for comparison</h2>
<form onchange="notify(event)" oninput="notify(event)">
  <fieldset>
    <gux-form-field resize="none" label-position="above">
      <textarea
        slot="input"
        name="textarea"
        placeholder="Type here..."
      ></textarea>
      <label slot="label">No Resize</label>
    </gux-form-field>

    <gux-form-field resize="manual" label-position="above">
      <textarea
        slot="input"
        name="textarea"
        placeholder="Type here..."
      ></textarea>
      <label slot="label">Manual Resize</label>
    </gux-form-field>

    <gux-form-field resize="auto" label-position="above">
      <textarea
        slot="input"
        name="textarea"
        placeholder="Type here..."
      ></textarea>
      <label slot="label">Auto Resize</label>
    </gux-form-field>

    <gux-form-field label-position="above">
      <textarea
        slot="input"
        name="textarea"
        placeholder="Type here..."
      ></textarea>
      <label slot="label">Text Area with Error Label</label>
      <span slot="error">Response must be less than 2000 characters</span>
    </gux-form-field>
  </fieldset>
</form>
