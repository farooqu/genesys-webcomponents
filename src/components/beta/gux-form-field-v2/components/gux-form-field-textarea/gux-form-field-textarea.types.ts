export type GuxFormFieldTextAreaResize = 'none' | 'manual' | 'auto';
