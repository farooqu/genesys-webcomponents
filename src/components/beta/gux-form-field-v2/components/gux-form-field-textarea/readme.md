# gux-form-field-textarea-beta



<!-- Auto Generated Below -->


## Properties

| Property        | Attribute        | Description | Type                                    | Default     |
| --------------- | ---------------- | ----------- | --------------------------------------- | ----------- |
| `labelPosition` | `label-position` |             | `"above" \| "beside" \| "screenreader"` | `undefined` |
| `resize`        | `resize`         |             | `"auto" \| "manual" \| "none"`          | `undefined` |


## Slots

| Slot      | Description                     |
| --------- | ------------------------------- |
| `"error"` | Optional slot for error message |
| `"input"` | Required slot for input tag     |
| `"label"` | Required slot for label tag     |


## Dependencies

### Depends on

- [gux-icon](../../../../stable/gux-icon)

### Graph
```mermaid
graph TD;
  gux-form-field-textarea-beta --> gux-icon
  style gux-form-field-textarea-beta fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
