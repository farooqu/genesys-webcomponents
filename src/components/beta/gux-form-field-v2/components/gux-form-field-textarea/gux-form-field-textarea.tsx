import { Component, Element, h, JSX, Prop, State } from '@stencil/core';

import { OnMutation } from '../../../../../utils/decorator/on-mutation';
import { onDisabledChange } from '../../../../../utils/dom/on-attribute-change';
import { onRequiredChange } from '../../../../../utils/dom/on-attribute-change';
import { preventBrowserValidationStyling } from '../../../../../utils/dom/prevent-browser-validation-styling';
import { trackComponent } from '../../../../../usage-tracking';

import { GuxFormFieldContainer } from '../../functional-components/gux-form-field-container/gux-form-field-container';
import { GuxFormFieldError } from '../../functional-components/gux-form-field-error/gux-form-field-error';
import { GuxFormFieldLabel } from '../../functional-components/gux-form-field-label/gux-form-field-label';

import { GuxFormFieldLabelPosition } from '../../gux-form-field.types';
import {
  hasErrorSlot,
  getComputedLabelPosition,
  validateFormIds
} from '../../gux-form-field.servce';

import { GuxFormFieldTextAreaResize } from './gux-form-field-textarea.types';
/**
 * @slot input - Required slot for input tag
 * @slot label - Required slot for label tag
 * @slot error - Optional slot for error message
 */
@Component({
  styleUrl: 'gux-form-field-textarea.less',
  tag: 'gux-form-field-textarea-beta',
  shadow: true
})
export class GuxFormFieldTextarea {
  private input: HTMLTextAreaElement;
  private label: HTMLLabelElement;
  private disabledObserver: MutationObserver;
  private requiredObserver: MutationObserver;
  private textareaContainerElement: HTMLDivElement;

  @Element()
  private root: HTMLElement;

  @Prop()
  resize: GuxFormFieldTextAreaResize;

  @Prop()
  labelPosition: GuxFormFieldLabelPosition;

  @State()
  private computedLabelPosition: GuxFormFieldLabelPosition = 'above';

  @State()
  private disabled: boolean;

  @State()
  private required: boolean = true;

  @State()
  private hasError: boolean = false;

  @OnMutation({ childList: true, subtree: true })
  onMutation(): void {
    this.hasError = hasErrorSlot(this.root);
  }

  componentWillLoad(): void {
    this.setInput();
    this.setLabel();

    this.hasError = hasErrorSlot(this.root);

    trackComponent(this.root, { variant: this.variant });
  }

  componentDidLoad(): void {
    this.updateHeight(this.textareaContainerElement, this.input, this.resize);
  }

  disconnectedCallback(): void {
    this.disabledObserver.disconnect();
    this.requiredObserver.disconnect();
  }

  render(): JSX.Element {
    return (
      <GuxFormFieldContainer labelPosition={this.computedLabelPosition}>
        <GuxFormFieldLabel
          position={this.computedLabelPosition}
          required={this.required}
        >
          <slot name="label" onSlotchange={() => this.setLabel()} />
        </GuxFormFieldLabel>
        <div class="gux-input-and-error-container">
          <div
            ref={el => (this.textareaContainerElement = el)}
            class={{
              'gux-input': true,
              [`gux-resize-${this.resize}`]: true,
              'gux-disabled': this.disabled,
              'gux-input-error': this.hasError
            }}
          >
            <slot name="input" />
          </div>
          <GuxFormFieldError hasError={this.hasError}>
            <slot name="error" />
          </GuxFormFieldError>
        </div>
      </GuxFormFieldContainer>
    ) as JSX.Element;
  }

  private get variant(): string {
    const labelPositionVariant = this.labelPosition
      ? this.labelPosition.toLowerCase()
      : 'none';

    return `${this.resize}-${labelPositionVariant}`;
  }

  private setInput(): void {
    this.input = this.root.querySelector('textarea[slot="input"]');

    preventBrowserValidationStyling(this.input);

    this.updateHeight(this.textareaContainerElement, this.input, this.resize);

    this.input.addEventListener('input', () => {
      this.updateHeight(this.textareaContainerElement, this.input, this.resize);
    });

    this.disabled = this.input.disabled;
    this.required = this.input.required;

    this.disabledObserver = onDisabledChange(
      this.input,
      (disabled: boolean) => {
        this.disabled = disabled;
      }
    );
    this.requiredObserver = onRequiredChange(
      this.input,
      (required: boolean) => {
        this.required = required;
      }
    );

    validateFormIds(this.root, this.input);
  }

  private setLabel(): void {
    this.label = this.root.querySelector('label[slot="label"]');

    this.computedLabelPosition = getComputedLabelPosition(
      this.label,
      this.labelPosition
    );
  }

  private updateHeight(
    container: HTMLElement,
    input: HTMLTextAreaElement,
    resize: GuxFormFieldTextAreaResize
  ): void {
    if (resize === 'auto') {
      if (container) {
        container.dataset.replicatedValue = input.value;
        container.style.maxHeight = input.style.maxHeight;
      }
    }
  }
}
