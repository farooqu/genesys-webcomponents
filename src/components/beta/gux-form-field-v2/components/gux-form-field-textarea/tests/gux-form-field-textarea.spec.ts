import { newSpecPage } from '@stencil/core/testing';
import MutationObserver from 'mutation-observer';

import { GuxFormFieldTextarea } from '../gux-form-field-textarea';

const components = [GuxFormFieldTextarea];
const language = 'en';

describe('gux-form-field-textarea-beta', () => {
  beforeEach(async () => {
    (
      global as NodeJS.Global & {
        MutationObserver: any;
      }
    ).MutationObserver = MutationObserver;
  });

  it('should build', async () => {
    const html = `
      <gux-form-field-textarea-beta>
        <textarea slot="input"></textarea>
        <label slot="label">Label</label>
        <span slot="error">Error message</span>
      </gux-form-field-textarea-beta>
    `;
    const page = await newSpecPage({ components, html, language });

    expect(page.rootInstance).toBeInstanceOf(GuxFormFieldTextarea);
  });

  describe('#render', () => {
    describe('resize', () => {
      ['', 'resize="auto"', 'resize="manual"', 'resize="none"'].forEach(
        (componentAttribute, index) => {
          it(`should render component as expected (${index + 1})`, async () => {
            const html = `
            <gux-form-field-textarea-beta ${componentAttribute}>
              <textarea slot="input"></textarea>
              <label slot="label">Label</label>
            </gux-form-field-textarea-beta>
          `;
            const page = await newSpecPage({ components, html, language });

            expect(page.root).toMatchSnapshot();
          });
        }
      );
    });

    describe('label-position', () => {
      [
        '',
        'label-position="above"',
        'label-position="beside"',
        'label-position="screenreader"'
      ].forEach((componentAttribute, index) => {
        it(`should render component as expected (${index + 1})`, async () => {
          const html = `
            <gux-form-field-textarea-beta ${componentAttribute}>
              <textarea slot="input"></textarea>
              <label slot="label">Label</label>
            </gux-form-field-textarea-beta>
          `;
          const page = await newSpecPage({ components, html, language });

          expect(page.root).toMatchSnapshot();
        });
      });
    });

    describe('input attributes', () => {
      ['', 'disabled', 'required'].forEach((inputAttribute, index) => {
        it(`should render component as expected (${index + 1})`, async () => {
          const html = `
            <gux-form-field-textarea-beta>
              <textarea slot="input" ${inputAttribute}></textarea>
              <label slot="label">Label</label>
            </gux-form-field-textarea-beta>
          `;
          const page = await newSpecPage({ components, html, language });

          expect(page.root).toMatchSnapshot();
        });
      });
    });
  });
});
