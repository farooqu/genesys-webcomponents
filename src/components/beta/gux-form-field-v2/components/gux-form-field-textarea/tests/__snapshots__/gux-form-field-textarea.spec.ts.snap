// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-form-field-textarea-beta #render input attributes should render component as expected (1) 1`] = `
<gux-form-field-textarea-beta>
  <mock:shadow-root>
    <div class="gux-above gux-form-field-container">
      <div class="gux-above gux-form-field-label">
        <slot name="label"></slot>
      </div>
      <div class="gux-input-and-error-container">
        <div class="gux-input gux-resize-undefined">
          <slot name="input"></slot>
        </div>
        <div class="gux-form-field-error">
          <gux-icon decorative="" icon-name="alert-warning-octogon"></gux-icon>
          <div class="gux-message">
            <slot name="error"></slot>
          </div>
        </div>
      </div>
    </div>
  </mock:shadow-root><textarea id="gux-form-field-input-i" slot="input"></textarea>
  <label for="gux-form-field-input-i" slot="label">
    Label
  </label>
</gux-form-field-textarea-beta>
`;

exports[`gux-form-field-textarea-beta #render input attributes should render component as expected (2) 1`] = `
<gux-form-field-textarea-beta>
  <mock:shadow-root>
    <div class="gux-above gux-form-field-container">
      <div class="gux-above gux-form-field-label">
        <slot name="label"></slot>
      </div>
      <div class="gux-input-and-error-container">
        <div class="gux-input gux-resize-undefined">
          <slot name="input"></slot>
        </div>
        <div class="gux-form-field-error">
          <gux-icon decorative="" icon-name="alert-warning-octogon"></gux-icon>
          <div class="gux-message">
            <slot name="error"></slot>
          </div>
        </div>
      </div>
    </div>
  </mock:shadow-root><textarea disabled="" id="gux-form-field-input-i" slot="input"></textarea>
  <label for="gux-form-field-input-i" slot="label">
    Label
  </label>
</gux-form-field-textarea-beta>
`;

exports[`gux-form-field-textarea-beta #render input attributes should render component as expected (3) 1`] = `
<gux-form-field-textarea-beta>
  <mock:shadow-root>
    <div class="gux-above gux-form-field-container">
      <div class="gux-above gux-form-field-label">
        <slot name="label"></slot>
      </div>
      <div class="gux-input-and-error-container">
        <div class="gux-input gux-resize-undefined">
          <slot name="input"></slot>
        </div>
        <div class="gux-form-field-error">
          <gux-icon decorative="" icon-name="alert-warning-octogon"></gux-icon>
          <div class="gux-message">
            <slot name="error"></slot>
          </div>
        </div>
      </div>
    </div>
  </mock:shadow-root><textarea id="gux-form-field-input-i" required="" slot="input"></textarea>
  <label for="gux-form-field-input-i" slot="label">
    Label
  </label>
</gux-form-field-textarea-beta>
`;

exports[`gux-form-field-textarea-beta #render label-position should render component as expected (1) 1`] = `
<gux-form-field-textarea-beta>
  <mock:shadow-root>
    <div class="gux-above gux-form-field-container">
      <div class="gux-above gux-form-field-label">
        <slot name="label"></slot>
      </div>
      <div class="gux-input-and-error-container">
        <div class="gux-input gux-resize-undefined">
          <slot name="input"></slot>
        </div>
        <div class="gux-form-field-error">
          <gux-icon decorative="" icon-name="alert-warning-octogon"></gux-icon>
          <div class="gux-message">
            <slot name="error"></slot>
          </div>
        </div>
      </div>
    </div>
  </mock:shadow-root><textarea id="gux-form-field-input-i" slot="input"></textarea>
  <label for="gux-form-field-input-i" slot="label">
    Label
  </label>
</gux-form-field-textarea-beta>
`;

exports[`gux-form-field-textarea-beta #render label-position should render component as expected (2) 1`] = `
<gux-form-field-textarea-beta label-position="above">
  <mock:shadow-root>
    <div class="gux-above gux-form-field-container">
      <div class="gux-above gux-form-field-label">
        <slot name="label"></slot>
      </div>
      <div class="gux-input-and-error-container">
        <div class="gux-input gux-resize-undefined">
          <slot name="input"></slot>
        </div>
        <div class="gux-form-field-error">
          <gux-icon decorative="" icon-name="alert-warning-octogon"></gux-icon>
          <div class="gux-message">
            <slot name="error"></slot>
          </div>
        </div>
      </div>
    </div>
  </mock:shadow-root><textarea id="gux-form-field-input-i" slot="input"></textarea>
  <label for="gux-form-field-input-i" slot="label">
    Label
  </label>
</gux-form-field-textarea-beta>
`;

exports[`gux-form-field-textarea-beta #render label-position should render component as expected (3) 1`] = `
<gux-form-field-textarea-beta label-position="beside">
  <mock:shadow-root>
    <div class="gux-beside gux-form-field-container">
      <div class="gux-beside gux-form-field-label">
        <slot name="label"></slot>
      </div>
      <div class="gux-input-and-error-container">
        <div class="gux-input gux-resize-undefined">
          <slot name="input"></slot>
        </div>
        <div class="gux-form-field-error">
          <gux-icon decorative="" icon-name="alert-warning-octogon"></gux-icon>
          <div class="gux-message">
            <slot name="error"></slot>
          </div>
        </div>
      </div>
    </div>
  </mock:shadow-root><textarea id="gux-form-field-input-i" slot="input"></textarea>
  <label for="gux-form-field-input-i" slot="label">
    Label
  </label>
</gux-form-field-textarea-beta>
`;

exports[`gux-form-field-textarea-beta #render label-position should render component as expected (4) 1`] = `
<gux-form-field-textarea-beta label-position="screenreader">
  <mock:shadow-root>
    <div class="gux-form-field-container gux-screenreader">
      <div class="gux-form-field-label gux-screenreader">
        <slot name="label"></slot>
      </div>
      <div class="gux-input-and-error-container">
        <div class="gux-input gux-resize-undefined">
          <slot name="input"></slot>
        </div>
        <div class="gux-form-field-error">
          <gux-icon decorative="" icon-name="alert-warning-octogon"></gux-icon>
          <div class="gux-message">
            <slot name="error"></slot>
          </div>
        </div>
      </div>
    </div>
  </mock:shadow-root><textarea id="gux-form-field-input-i" slot="input"></textarea>
  <label for="gux-form-field-input-i" slot="label">
    Label
  </label>
</gux-form-field-textarea-beta>
`;

exports[`gux-form-field-textarea-beta #render resize should render component as expected (1) 1`] = `
<gux-form-field-textarea-beta>
  <mock:shadow-root>
    <div class="gux-above gux-form-field-container">
      <div class="gux-above gux-form-field-label">
        <slot name="label"></slot>
      </div>
      <div class="gux-input-and-error-container">
        <div class="gux-input gux-resize-undefined">
          <slot name="input"></slot>
        </div>
        <div class="gux-form-field-error">
          <gux-icon decorative="" icon-name="alert-warning-octogon"></gux-icon>
          <div class="gux-message">
            <slot name="error"></slot>
          </div>
        </div>
      </div>
    </div>
  </mock:shadow-root><textarea id="gux-form-field-input-i" slot="input"></textarea>
  <label for="gux-form-field-input-i" slot="label">
    Label
  </label>
</gux-form-field-textarea-beta>
`;

exports[`gux-form-field-textarea-beta #render resize should render component as expected (2) 1`] = `
<gux-form-field-textarea-beta resize="auto">
  <mock:shadow-root>
    <div class="gux-above gux-form-field-container">
      <div class="gux-above gux-form-field-label">
        <slot name="label"></slot>
      </div>
      <div class="gux-input-and-error-container">
        <div class="gux-input gux-resize-auto" data-replicated-value="undefined">
          <slot name="input"></slot>
        </div>
        <div class="gux-form-field-error">
          <gux-icon decorative="" icon-name="alert-warning-octogon"></gux-icon>
          <div class="gux-message">
            <slot name="error"></slot>
          </div>
        </div>
      </div>
    </div>
  </mock:shadow-root><textarea id="gux-form-field-input-i" slot="input"></textarea>
  <label for="gux-form-field-input-i" slot="label">
    Label
  </label>
</gux-form-field-textarea-beta>
`;

exports[`gux-form-field-textarea-beta #render resize should render component as expected (3) 1`] = `
<gux-form-field-textarea-beta resize="manual">
  <mock:shadow-root>
    <div class="gux-above gux-form-field-container">
      <div class="gux-above gux-form-field-label">
        <slot name="label"></slot>
      </div>
      <div class="gux-input-and-error-container">
        <div class="gux-input gux-resize-manual">
          <slot name="input"></slot>
        </div>
        <div class="gux-form-field-error">
          <gux-icon decorative="" icon-name="alert-warning-octogon"></gux-icon>
          <div class="gux-message">
            <slot name="error"></slot>
          </div>
        </div>
      </div>
    </div>
  </mock:shadow-root><textarea id="gux-form-field-input-i" slot="input"></textarea>
  <label for="gux-form-field-input-i" slot="label">
    Label
  </label>
</gux-form-field-textarea-beta>
`;

exports[`gux-form-field-textarea-beta #render resize should render component as expected (4) 1`] = `
<gux-form-field-textarea-beta resize="none">
  <mock:shadow-root>
    <div class="gux-above gux-form-field-container">
      <div class="gux-above gux-form-field-label">
        <slot name="label"></slot>
      </div>
      <div class="gux-input-and-error-container">
        <div class="gux-input gux-resize-none">
          <slot name="input"></slot>
        </div>
        <div class="gux-form-field-error">
          <gux-icon decorative="" icon-name="alert-warning-octogon"></gux-icon>
          <div class="gux-message">
            <slot name="error"></slot>
          </div>
        </div>
      </div>
    </div>
  </mock:shadow-root><textarea id="gux-form-field-input-i" slot="input"></textarea>
  <label for="gux-form-field-input-i" slot="label">
    Label
  </label>
</gux-form-field-textarea-beta>
`;
